[47519] Single-date mode for work package date pickers

HDinger opened the pull request:
# Ticket
https://community.openproject.org/projects/openproject/work_packages/47519/activity

⚠️ By now, this also contains https://community.openproject.org/projects/openproject/work_packages/59540/activity, which was already reviewed and approved ⚠️ 

# What are you trying to accomplish?
Create a "single date mode" for the new datepicker

## Screenshots

![Feb-19-2025 12-05-03](https://github.com/user-attachments/assets/077e16ec-7fd1-404e-a0a6-6312e12169b1)


# ToDo

- [x] Show buttons instead of fields when start date OR due date are empty
- [x] Clicking the button, should reveal the fields
- [x] Once clicked, the button will not appear again as long as the modal is opened
  - [x] Catch turbo frame updates
  - [x] Catch morph updats
- [x] Entering a duration will automatically switch from single to range mode
- [x] Adapt tests to the new behaviour

HDinger committed:
First draft of a single date mode in datepicker. Therefore, the form was manually build to add buttons into it

HDinger committed:
Move controller to the top so that the toggled state is preserved after a turbo frame re-rendering

HDinger committed:
Exclude the Datebuttons from morphing to avoid that they are shown every time the form udpates itself

HDinger committed:
When duration is updated, change to range mode

HDinger committed:
Add toggling behaviour for single date mode

HDinger committed:
Correctly focus the next date in single date mode

HDinger committed:
Adapt tests to new single-date mode

HDinger committed:
Fix some minor spacing issues

HDinger committed:
Make Milestones work correctly again

HDinger committed:
Adapt tests to new single-date mode

bsatarnejad committed:
change last to closest

HDinger committed:
Check for other fields values to know when to toggle the visibility automatically (and when not)

HDinger committed:
Move toggling of date mode to the backend instead of trying to remember the sate in the frontend

HDinger committed:
Merge pull request #17948 from opf/61448-replace-last-by-closest-in-the-no-predecessors-placeholder-in-date-picker

[61448] Replace "last" by "closest" in the "no predecessors" placeholder in date picker

HDinger committed:
Pass the "date_mode" param correctly

HDinger committed:
Some mobile refinements

HDinger committed:
Set focus when activating a date field and expect  that in a test

HDinger committed:
Hide clear button when the field is disabled

HDinger committed:
Read the values correclty after a refresh. For whatever reason the stimulus targets still reference the old fields and thus the old values

HDinger committed:
Correct the logic for displaying the clear button

HDinger committed:
Show a different path when we are in the create case

HDinger committed:
Use the passed value instead of trying to get the information from the params

HDinger committed:
Correctly read out values everytime we change the focus. Otherwise we sometimes get to an old data state

HDinger committed:
Small code improvements

HDinger committed:
Let progress/preview controller inherit from the generic workpackage/dialog/preview controller again while keeping the special behaviour of focus/blur events.

HDinger committed:
Pass `date_mode` from existing parameter into the component instead of trying to recalculate it again

HDinger committed:
Focus the correct field in single date mode when both fields are empty

HDinger committed:
Add test cases for single-date mode

HDinger committed:
Do not switch back to hiding fields when the field was already touched

psatyal commented:
Hi @HDinger,

This is tremendous work! Very glad to be able to test it. It looks like the logic and the interaction works well but there are just a few visual kinks that need ironing out. I've listed them below:

**Inconsistent width and spacing**
The width of the datepicker seems to vary according to different selections: for example, when switching between Manual and Automatic and then the Predecessor tab, and it looks like this is mostly due to the width of the inputs with adds additional padding. The Predecessor has the correct padding. Here's a video:

https://github.com/user-attachments/assets/8d73398e-a9bd-4112-af82-fd501ada4abf

For these different widths, I've add guides that show deviations (red) from the correct padding (green):

**Single mode**
![Single-manually-scheduled-padding-incorrect](https://github.com/user-attachments/assets/46f673b8-8137-4f86-9440-8a1b03b68b27)  

We always had a bit of extra padding around the Flatpickr mini-calendars, but not as much as this (although there seems to be some extra padding above the mini-calendar):

![Current-datepicker-community-padding](https://github.com/user-attachments/assets/4b8d6012-e541-495f-8f06-4d0fc00a7599)


**Range mode**
The paddings waay too big and are different between auto and modes modes.

![Range-auto-scheduled-padding-incorrect](https://github.com/user-attachments/assets/94c0f1f3-f776-4a9f-b860-8f1df99d62e2)

![Range-manually-scheduled-padding-incorrect](https://github.com/user-attachments/assets/46e27237-7468-4213-9386-aea51f22bd2e)

Apart from these issues with padding, I came across an odd error when I removed the dates from a work package or went to single mode (removed one date) and tried to edit it again:

https://github.com/user-attachments/assets/a10e13d3-0457-412f-b17b-0e8ca6ac36d8

I'll ignore issues with focus for now (as requested) and I'll continue to test for other visual quirks, but I think we can address these first.

Thank you!

Cheers
Parimal

myabc committed:
Implement RelationGroup#closest_relation

Adds `WorkPackageRelationsTab::ClosestRelation`, encapsulating the
sorting logic in a custom comparator.

myabc committed:
[#59540] Highlight closest relation in relations tab

Updates `RelationComponent#initialize` to accept a `closest:` argument.

myabc committed:
Expect closest in Primerized relations tab feature

myabc committed:
Exclude relations w/o dates in #closest_relations

HDinger committed:
Keep focus on the active field when deleting it's value

HDinger committed:
Please rubocop and fix last check for showing the text field in single date mode

HDinger committed:
Fix some spacing issues

myabc committed:
DRY up RelationsMediator type predicates

Makes uses of `ActiveSupport::StringInquirer`.

HDinger committed:
Handle special cases that two fields have been cleared in which we assume that the user wants to delete all three values

HDinger committed:
Add special behaviour when clearing a date and setting the duration afterwards

myabc committed:
Remove unused RelationsMediator#closest_relations

This method was added anticipating possible changes to the sort order of
follows relations in the UI. We have now determined that we will keep
the current behaviour (i.e. sorting by `WorkPackage#created_at`).

myabc committed:
Further flesh out Primerized relations tab feature

Adds expectations for "not closest" predecessor.

myabc committed:
Add Capybara selector, matchers for Primer Label

myabc committed:
Teach Label selector scheme: expression_filter

cbliard committed:
Merge pull request #17942 from opf/feature/59540-highlight-closest-predecessor-in-relations-tab

[#59540] highlight closest predecessor in relations tab

cbliard committed:
Merge branch 'feature/42388-new-automatic-scheduling-mode' into feature/47519-single-date-mode-for-work-package-date-pickers

HDinger committed:
Re-add styles for "modal-single-date-picker" component which went missing when reworking the datepicker styles. Thereby the bulk edit form was changed to use the basic single date picker for CustomFields which maps to the component that is used in the single view. It made no sense to use the modal variant here.

myabc committed:
Remove unnecessary #lazy call in #closest_relation

HDinger committed:
Unset the height and width restrictions for mobile

ulferts reviewed (changes requested):
Hi @HDinger , I haven't completed the review yet but thought it makes sense to already deliver a first chunk with behavioural shortcomings discovered while playing around with the functionality.

More might follow by working through the code.

### Selected dates get cleared when changing/deleting the duration

* [x] Selected dates get cleared when changing/deleting the duration

![Kapture 2025-03-03 at 11 55 21](https://github.com/user-attachments/assets/c5eceba9-5dc0-4bae-8662-2ff0725404de)
 
  * [x] The same happens if the duration is removed and not changed
  * [x] ~This also shows a visual glitch in that the days between the start and the finish dates are not colored (light blue) when the duration is input first. Only the start/finish date are colored (dark blue)~ Tracked in https://community.openproject.org/wp/61957

### Modifying the duration and then selecting a date

This one is probably not a bug as such and seems to rely on the behaviour that is also present in the range mode. But it does feel even weirder in the single date mode.

![Kapture 2025-03-03 at 12 04 44](https://github.com/user-attachments/assets/77c0fb55-e099-491f-adea-a9323bdad429)

In case only the finish date is active, I insert a duration, then select a date, I would have expected that date to be the finish date.

The behaviour as such is consistent with the one in range mode. If I insert a duration there and then select a date, the selected date turns out to be the start date as well. 

What would have helped in those cases IMO is turning the start button into a field after the duration is set and the mouse is moved into the date picker. This contradicts the requirements though. IMO, the focus should switch to the start input field in the range mode as well.

A similar behaviour is also present when:
* Having a start date
* Adding a duration
* Select a date via the date picker that is after the current finish dates. 
=> The start date is set to the new date and the rest of the information is removed. 

Focusing the start field when moving moving the mouse into the range would help there as well.

### Involuntarily switching to range mode

* [x] Involuntarily switching to range mode

![Kapture 2025-03-03 at 12 56 13](https://github.com/user-attachments/assets/4c6f76e9-42cc-4ee3-b6b4-2c292aafc01b)

### Triggering the date picker in single mode from the work package table

* [x] Triggering the date picker in single mode from the work package table
 
![Kapture 2025-03-03 at 13 40 23](https://github.com/user-attachments/assets/87a305ce-966c-46f7-84b5-6d1049715c72)

### Nitpicking: Segmented control flickering when input field is cleared

![Kapture 2025-03-03 at 13 55 35](https://github.com/user-attachments/assets/6fd25cdb-ae2a-44e3-a4de-699e9c8e1dff)

- [x] Tracked in https://community.openproject.org/work_packages/61959

ulferts commented on spec/features/work_packages/datepicker/datepicker_single_date_mode_logic_spec.rb:
```diff
@@ -0,0 +1,788 @@
+require "spec_helper"
+require "support/edit_fields/edit_field"
+
+RSpec.describe "Datepicker: Single-date mode logic test cases (WP #61146)", :js, with_settings: { date_format: "%Y-%m-%d" } do
```
I don't understand the separation of the tests in this file here and the ones in `datepicker_logic_spec.rb`. There are test cases in the mentioned file, that also start in the single mode now. 

Wouldn't it make more sense to have them all in one file to ensure that they are consolidated. And the date picker switches between the modes easily on user interaction. If that is not wanted however, then I would split the test cases more strictly between the two files and rename the one that now has the range date mode.

HDinger replied:
The separation of these scenarios was done by the product team. The test files are a one to one representation of the belonging test scenario tickets, in that case https://community.openproject.org/work_packages/61146 . The test cases describedn in the `datepicker_logic_spec` are the representatives of https://community.openproject.org/work_packages/60670.

ulferts commented on spec/features/work_packages/datepicker/datepicker_single_date_mode_logic_spec.rb:
```diff
@@ -0,0 +1,788 @@
+
+  before do
+    Setting.available_languages << current_user.language
+    I18n.locale = current_user.language
```
I assume those are not needed here.

ulferts commented on app/components/work_packages/date_picker/date_form_component.html.erb:
```diff
@@ -0,0 +1,47 @@
+                                               classes: "wp-datepicker-dialog-date-form--add-button",
+                                               test_selector: "wp-datepicker--show-due-date")) do |button|
+              button.with_leading_visual_icon(icon: :plus)
+              I18n.t("attributes.due_date")
```
I personally would prefer using `WorkPackage.human_attribute_name('due_date')` as it is less dependent on the translation as more places are looked at.

ulferts commented on app/components/work_packages/date_picker/date_form_component.html.erb:
```diff
@@ -0,0 +1,47 @@
+      end
+
+      body.with_column(classes: "wp-datepicker-dialog-date-form--duration") do
+        render(Primer::Alpha::TextField.new(**text_field_options(name: :duration, label: I18n.t("activerecord.attributes.work_package.duration"))))
```
`WorkPackage.human_attribute_name('duration')` works here as well :)

ulferts commented on app/components/work_packages/date_picker/date_form_component.rb:
```diff
@@ -200,13 +176,43 @@ def default_field_options(name)
       end
+
+      def single_date_field_button_link(focused_field)
+        permitted_params = params.merge(date_mode: "range", focused_field:).permit!
```
I don't think that the places I have seen pose an immediate security risk. However, it would be cleaner if this functionality, since it is used in a number of places was to be extracted to some shared method which could then also permit the parameters on a per need basis.

ulferts commented on app/components/work_packages/date_picker/form_content_component.html.erb:
```diff
@@ -12,61 +12,89 @@
+                    },
+                    test_selector: "op-datepicker-modal--scheduling_manual",
+                    label: I18n.t("work_packages.datepicker_modal.mode.manual"),
+                    title: I18n.t("work_packages.datepicker_modal.mode.manual"),
```
This has been in there before. But what is the reason for displaying a title that has the same string as the label. To me this seems to be noise.

HDinger committed:
Show the datepicker in single date mode when being opened from the Table

ulferts commented on app/components/work_packages/date_picker/form_content_component.html.erb:
```diff
@@ -12,61 +12,89 @@
+                      qa_selected: !schedule_manually
+                    },
+                    test_selector: "op-datepicker-modal--scheduling_automatic",
+                    disabled: work_package.new_record?,
```
This is the line of code that leads to the behaviour of blocking the selection for unpersisted work packages we talked about, @HDinger. I did not try removing it, though.

HDinger replied:
Removing it will crash the form as the current implementation tries to find a WP with id which is not present yet. I created a ticket for that: https://community.openproject.org/wp/61960

HDinger committed:
Preserve the field value of only one field was touched anyway

HDinger committed:
Always preserve the range mode once it is entered

ulferts commented on app/controllers/work_packages/date_picker_controller.rb:
```diff
@@ -145,18 +152,23 @@ def datepicker_modal_component
     when "work_package[duration]"
       :duration
+    when "work_package[due_date]"
+      handle_focus_order_for_fields(:due_date, :start_date)
```
This controller does a lot of work that IMO should not be in it. 

The normal work routine is something like:
* Get the params and clean them up.
* Modify the a backend object.
* Render the response.

In this case there is no backend object. But then again, it is modifying an object which could be described as the state. It is not persisted of course but it doesn't have to be to allow the pieces to fall into place again.

What do you think? Could it be an option of introducing a `DatePickerState` (name open for discussions) object that takes the parameters, calculates on them (e.g. which field to focus) and is then passed into the component for rendering? It would take the params and then have methods for querying on the current state. Probably, most of the methods in this controller could just be moved into that object without a lot of hassle. 

Further down, in `handle_form_cleared` the params are also manipulated. This to me is another indicator that some state object is necessary for cleaning up.

HDinger replied:
I tracked that in https://community.openproject.org/wp/61966

HDinger committed:
With duration in focus, the datepicker is clicked. In that case we have to decide wich field the value is for. The due date should be seleected when it is empty AND (the start date is either filled OR not touched). The later is basically the case when the datepicker is in single date mode

ulferts commented on frontend/src/stimulus/controllers/dynamic/work-packages/progress/preview.controller.ts:
```diff
@@ -28,64 +28,35 @@
 
-    this.progressInputTargets.forEach((target) => {
+  registerFieldInputListeners() {
+    this.fieldInputTargets.forEach((target) => {
```
I think this can be simplified using the [stimulus target connect/disconnect callbacks](https://stimulus.hotwired.dev/reference/targets#connected-and-disconnected-callbacks). Using those, I don't see a reason to define and mange those listeners in the superclass at all.

ulferts replied:
I might be missing parts but I don't quite get why we don't set the [actions via stimulus directly](https://stimulus.hotwired.dev/reference/actions). To me, this seems to be an indirection.

HDinger replied:
Done 👍

ulferts reviewed (changes requested):
Hi @HDinger , I now managed to work through the code. There are a couple of remarks. Not all of them are a necessity to have fixed. 

From the bugs in my first review, those that needed to be fixed are already fixed, so that is already looking good.

psatyal commented:
Hi @HDinger,

Thanks for your work on this. Here are few other issues, other then ones @ulferts already reported:

**Clear icon:**
![datepicker-clear-icon-too-prominent](https://github.com/user-attachments/assets/4c6599b1-0fdc-4ac1-bbe5-290ebe1e775c)
- [x] ~Clear icon on the date fields is too prominent. Let's try using the `x` instead of `x-circle-fill`~ (HD: we can't influence the icon, as it is part of the Primer TextInput component and not changeable) 

**Banner:** 
<img width="632" alt="Banner padding and centering" src="https://github.com/user-attachments/assets/ea0f8f34-14fb-41ce-b1e2-e8f89f5ce26a" />
- [x] ~The top/bottom margins are too large. Possible fix: `var(--base-size-4)` instead of `var(--base-size-8)`.~
- [x] ~The info icon isn't aligned with the text. We can either vertically center-align it, but even better would be to align it with the first line. Possible fix: `padding: var(--base-size-6) var(--base-size-8)` with `padding: var(--base-size-8) var(--
base-size-8)`.~ 
(HD) This is the standard Primer Banner. We will not change that.

**Does not auto-switch to auto**
- [x] ~Making a WP a parent to existing WP does not automatically switch it to automatic. But setting a manually scheduled WP as follower *does* set it dates automatically. Possible explanation: @cbliard did say that there were certain cases where work packages don't automatically switch to automatic yet. This might be one of the cases.~ Tracked here: https://community.openproject.org/work_packages/61965

**Only left part of mini calendar used**
https://github.com/user-attachments/assets/30ef3bf3-862d-4cba-b1d8-b8045d63376d
- [x] ~When entering dates, only the first part of the full-calendar is used (for example, if you go from the month on the left to the one on the right, the first calendar changes instead of the selection going to the second calendar).~ Tracked in https://community.openproject.org/wp/61961



**Date picker not showing**
https://github.com/user-attachments/assets/46e63e0b-c4e1-40d9-bea9-c3a6abdfd101
- [x] The date picker doesn't trigger in some cases if the duration is empty (I tried adding dates, then removing them, then adding a duration only, then removing that too).

**Mobile (tested on Safari and Firefox on iOS):**
![mobile-datepicker-top-half-cut-off](https://github.com/user-attachments/assets/6b20b388-9b48-405f-941d-3ecae6e55269)

  - [x] Rendering is off; the top half is cut off. All the elements above the start date field are neither accessible nor visible, and scrolling doesn't help (the label for start field, 'Working days only' checkbox, mode selector, banners).
  - [x] Duration field shows a "normal" keyboard input, it should show the numbers one.
  - [x] ~Date field shows "normal" input; it should show the browser native one. (Or the numbers or symbols one if native is too complicated)~ Will be part of https://community.openproject.org/wp/61051
  
![mobile-datepicker-keyboard-date](https://github.com/user-attachments/assets/969b4cc9-cfb2-4eac-9d99-4420ddbfba5f)
![mobile-datepicker-keyboard-duration](https://github.com/user-attachments/assets/690a2d72-451f-4f63-be40-ec97a965b6cc)

There might be others too, will continue testing.

Cheers
Parimal

HDinger committed:
Correctly touch and untouch fields when duration is edited. Thereby it makes a differences whether the fields have been changed or deleted. Depending on that, the fields is either preserved or overwritten

HDinger committed:
Make sure that the datepicker is also opened when the duration field is activated twice

HDinger committed:
Make sure the datepicker is scrollable on mobile and add some review feedback

HDinger committed:
Fix broken specs

HDinger committed:
Extract the event listeners into stimulus action calls

HDinger committed:
Merge branch 'dev' into feature/47519-single-date-mode-for-work-package-date-pickers

HDinger committed:
Merge branch 'dev' into feature/47519-single-date-mode-for-work-package-date-pickers

HDinger committed:
Rename FormContentComponent to FormComponent and register change event on the NWD checkbox

ulferts committed:
remove pending as the spec is stable locally

HDinger committed:
Attempt to fix the spec

ulferts committed:
use decimal separator fit for en

ulferts committed:
Merge remote-tracking branch 'origin/dev' into feature/47519-single-date-mode-for-work-package-date-pickers

ulferts reviewed (approved):
👍
